source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,630 @@
+            AZLogDebug( ( "Returned from Methods Callback." ) );
+
+            xResult = eAzureIoTHubClientSuccess;
+        }
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Check/Process messages for incoming device twin messages.
+ *
+ * */
+static uint32_t prvAzureIoTHubClientDeviceTwinProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                       AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                       void * pvPublishInfo )
+{
+    AzureIoTHubClientResult_t xResult;
+    AzureIoTHubClientTwinResponse_t xTwinResponse = { 0 };
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pvPublishInfo;
+    az_result xCoreResult;
+    az_iot_hub_client_twin_response xOutRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pcTopicName, xMQTTPublishInfo->usTopicNameLength );
+    uint32_t ulRequestID = 0;
+
+    /* Failed means no topic match */
+    xCoreResult = az_iot_hub_client_twin_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                               xTopicSpan, &xOutRequest );
+
+    if( az_result_failed( xCoreResult ) )
+    {
+        AZLogWarn( ( "Twin topic parsing failed: 0x%08x", xCoreResult ) );
+        xResult = eAzureIoTHubClientTopicNoMatch;
+    }
+    else
+    {
+        /* Verify the received publish is for the feature we have subscribed to. */
+        AZLogDebug( ( "Twin topic: %.*s. with payload : %.*s.",
+                      xMQTTPublishInfo->usTopicNameLength,
+                      xMQTTPublishInfo->pcTopicName,
+                      xMQTTPublishInfo->xPayloadLength,
+                      xMQTTPublishInfo->pvPayload ) );
+
+        xResult = eAzureIoTHubClientSuccess;
+
+        if( pxContext->_internal.callbacks.xTwinCallback )
+        {
+            if( az_span_size( xOutRequest.request_id ) == 0 )
+            {
+                xTwinResponse.xMessageType = eAzureIoTHubTwinDesiredPropertyMessage;
+            }
+            else
+            {
+                if( az_result_succeeded( xCoreResult = az_span_atou32( xOutRequest.request_id, &ulRequestID ) ) )
+                {
+                    if( ulRequestID & 0x01 )
+                    {
+                        xTwinResponse.xMessageType = eAzureIoTHubTwinReportedResponseMessage;
+                    }
+                    else
+                    {
+                        xTwinResponse.xMessageType = eAzureIoTHubTwinGetMessage;
+                    }
+                }
+                else
+                {
+                    /* Failed to parse the message */
+                    AZLogError( ( "Request ID parsing failed: 0x%08x", xCoreResult ) );
+                    xResult = eAzureIoTHubClientFailed;
```

Comment:
I understand that we need to set some signal if we get here so as not to invoke the callback.  However we don't want to return this to the caller in the PUBLISH processing logic, because it will interpret that as a TopicNotMatch and continue through other topics where really caller should fail.

Reply:
yea this is interesting. I worry that some error in a process function for a feature it isn't for could prevent it from getting processed by the correct function if we bail out on error. Not sure how recoverable we want this to be.

Reply:
@danewalton - tagging just since this got merged (congrats!).

I actually don't think this function should be returning an ERROR.  The caller doesn't care whether there was an error here or not.

This function (and friends) should be returning like a `bool` for "I claim ownership of this topic" or "I matched this topic."  And from caller perspective it's basically saying "OK, the topic matched, stop going through loop."  No matter what happens after the topic match, since we can only match one topic the caller should stop processing.